### ISSUE TITLE
Delete previously ingested NGAP Files

### ISSUE THREAD

**Issue by krisstanton** [NASA-IMPACT/csdap-cumulus#347]
As a result of the recent dataset migration tasks, we currently have a large amount of data that exists in multiple locations.
This task is to remove the granules and files that have already been migrated to the new CBA NGAP accounts.
The granules and files we are removing are the ones that are currently stored on the older NGAP system which is no longer referenced and now unused.
Note: Not all of the data has been migrated away from NGAP at this time so we need a process to carefully remove the files and granules that are already removed.

The way this works is, we get a list of ingested granules from EarthData for any given dataset
Note: This list will include granules that have been ingested AND granules that have been migrated.
If possible (if the CMR metadata allows) we need to check the pointer location of the granules from the CMR.  If the location points to DNS `csdap`, than that is a granule which has been fully migrated and published and lives in the NEW CBA system.  If the DNS points to DNS `csda`, then that is a granule which has been ingested but still lives in the OLD NGAP system.
We need to store which system we have identified it in and the datetime that it was checked.

The next step is to compare that list of `csdap` pointed granules against all the granules and files found in the new CBA system.
If a file is found in the new CBA system (the PROD account, buckets Protected and Public), we need to mark that it was found there and note the CBA Bucket and CBA Key path and datetime that it was observed (aws s3 list commands)

DRAFT Steps For the Process - Check off each of these as the process as been developed.
// Do each of these steps per dataset - the earthdata query, the CBA checkpoint, gathering the list of old NGAP bucket and Key addresses
- [x] Complete Ticket #341 
- - [x] Check CMR for existence of granules (make the master list from this query)
- - [x] Handle Pagination on the Query
- - [x] Output will be granule level.  Note that we will be getting mixed out from CMR (some granules from old NGAP and some from new CBA), the way to tell the difference is to examine their DNS pointer address on the file download `csdap` means CBA, `csda` means old NGAP - Helper to API Docs  https://cmr.earthdata.nasa.gov/search/site/docs/search/api.html    (UPDATE: Output is actually File Level, and are mining the S3 URLs from each result - Also using the S3 URL (bucket name) to determine if the granule is from `OLD_NGAP` or `NEW_CBA`.  The list we want to keep from the CMR is the list of anything found in `NEW_CBA` - which should be the vast majority of the URLs)
- [x] Check the CBA PROD account for files from each granule (that is confirmed found in EarthData to be from `csdap`. Update, the S3 URLs would not be published unless they are already successfully passed the `move` step during ingest, so this verification is actually now part of the CMR request script)
- - (UPDATE: Some of the older below items are actually handled within the CMR Requests script)
- - Make a list of every file (maybe aws cli s3 ls commands for this?)
- - Match each granule to each list of files (and their locations, both in Public and Protected)
- - Output a list of files to delete (CSV)
- [ ] Check the old NGAP PROD account to verify the key paths of each file to delete
- - [ ] Output a list of files to delete (CSV) - this list is the verified list for deleting files. 

Update (from the comment below): 
There are two methods we can use to delete files.
- `aws s3 rm <s3_uri>`
- `aws s3api delete-objects <params including bucket and keypath>`
We learned that with method 2, we can pass pretty much any key into the csv file for deletion.
Here are the observed outcomes (the bucket name is passed in with the s3api call)
- If the Key Exists, it gets deleted
- If the Key does not exist, nothing happens and this item is skipped - there is no error, in fact, the output looks the same as if there was a key that was deleted.
- If the Key is a duplicate or was already deleted, same effect as if it does not exist.
The big picture here is that the only reliable method we really have for confirming a delete is to check the metrics before and after we run the delete code.  
- [x] Batch Delete
- - [x] Find out what format we need for the batch delete (CSV, BucketName?, Keypath?, etc).
- - [x] Set up a test version of the batch delete (3 granules and their files) In the CBA Sandbox
- - - [x] New Bucket for testing a batch delete, test_CSV file pointing at only some of the files in those 3 granules.
- - - [x] Verify that the batch delete only deleted the items from the test_CSV file.
- - [x] Actually do the batch delete from NGAP

Perform all of the above steps per dataset - Check off each dataset as the deletions have been completed.
Here is the progress checklist on a per item basis
- [x] WV04_Pan_L1B
- [x] WV04_MSI_L1B
- [x] WV03_Pan_L1B
- [x] WV03_MSI_L1B
- [x] WV02_MSI_L1B
- [x] WV01_Pan_L1B
- [x] GE01_MSI_L1B

Metrics
URLs for Metrics
- Old NGAP: csdap-cumulus-prod-protected - URL: https://s3.console.aws.amazon.com/s3/buckets/csdap-cumulus-prod-protected?region=us-west-2&bucketType=general&tab=metrics
- Old NGAP: csdap-cumulus-prod-public - URL: https://s3.console.aws.amazon.com/s3/buckets/csdap-cumulus-prod-public?region=us-west-2&bucketType=general&tab=metrics

BEFORE Any Deletes
- Old NGAP: csdap-cumulus-prod-protected
  - Total Bucket Size: 	7.2 PB
  - Total Objects: 		123704894 	or 	(123,704,894) 	or ~124 million
- Old NGAP: csdap-cumulus-prod-public
  - Total Bucket Size: 	2.0 TB
  - Total Objects: 		13852198 	or 	(13,852,198)    	or ~14 million

WIP At about the halfway point (in number of files)
- Old NGAP: csdap-cumulus-prod-protected
  - Total Bucket Size: 2.1 PB
  - Total Objects: 62962962 or (62,962,962) or ~62 million
- Old NGAP: csdap-cumulus-prod-public
  - Total Bucket Size: 411.3 GB
  - Total Objects: 3700235 or (3,700,235) or ~3 million

AFTER Almost all of the Deletes
- Old NGAP: csdap-cumulus-prod-protected
  - Total Bucket Size: 1.1 PB
  - Total Objects: 43938644 or (43,938,644) or ~43 million
- Old NGAP: csdap-cumulus-prod-public
  - Total Bucket Size: 73.2 GB
  - Total Objects: 589943 or (589,943) or ~0.5 million

AFTER all of the Deletes (including duplicate 2-digit and 1-digit directory name bug)
- Placeholder

- [x] Remove the final duplicates (2 and 1 digit directory names) and update the metrics above again.
- [ ] Make an inventory so we know which granules were not fully ingested before shutting down NGAP server 
  - [x] Turning the inventory back on for Protected and Public buckets for 1 day AFTER the above item is completed 
  - [ ] Download the latest inventory (and place the info in MCP Delivery bucket)
  - [ ] Turn that inventory back off

**Comment by krisstanton** [comment]
WIP Update: 
The CMR requests script which gives us the large list of published S3 URLs is done and functioning.   
The script for combining any overlapping URLs from the output of the CMR requests script is done and functioning.
The script which reads the manifest from `OLD_NGAP` is mostly done.  
I am currently at the part where I'm coding in the compare (since the file paths might be different, I'm using the Granule ID to do the compare).
The remaining step after the manifest reading and compare script is to call the S3 Delete function against the CSV output from the Manifest reading and comparing script. 

Note: I also updated the description of this ticket above to more accurately reflect the process as it's being developed.

**Comment by krisstanton** [comment]
UPDATE: Here is an explicit listing of how our delete test should go

```
- Manually Create a bucket for testing the deleting of files.
- Manually upload 5 files with different names to it
- Manually create a CSV file which has two columns and 4 rows (Like this)
"bucket_name","key_for_file_1"
"bucket_name","key_for_file_2"
"bucket_name","key_for_file_2"
"bucket_name","key_for_file_that_does_not_exist"
- Trigger the delete 

- The Expected Test results should go like this:
- - File 1 should be deleted
- - File 2 should be deleted
- - File 3, 4, 5 should still exist in their buckets because we did not put a bucket in there yet
- - There might possibly be an error or log entry for 'key_for_file_2' since we duplicated that row (tried to delete it twice)
- - There might possibly be an error or log entry for 'key_for_file_that_does_not_exist' since that file does not exist.

- The idea is that we want to (a) know that the delete method works (b) know that the method WILL NOT delete files that we did not explicitly list and (c) as a bonus, find out if we can see any other behavior in the output logs (like files that did not exist or duplicates)
```

**Comment by krisstanton** [comment]
Update: 
There are two methods we can use to delete files.
- `aws s3 rm <s3_uri>`
- `aws s3api delete-objects <params including bucket and keypath>`
We learned that with method 2, we can pass pretty much any key into the csv file for deletion.
Here are the observed outcomes (the bucket name is passed in with the s3api call)
- If the Key Exists, it gets deleted
- If the Key does not exist, nothing happens and this item is skipped - there is no error, in fact, the output looks the same as if there was a key that was deleted.
- If the Key is a duplicate or was already deleted, same effect as if it does not exist.
The big picture here is that the only reliable method we really have for confirming a delete is to check the metrics before and after we run the delete code.  

Example Delete using the first method 
```
aws s3 rm s3://deletebucket-test-sbx/2018/356/WV04_20181222045557_73ba3830-9f58-4b42-9097-a9a1ff451090-inv_18DEC22045557-P1BS-059420282180_01_P005/WV04_20181222045557_73ba3830-9f58-4b42-9097-a9a1ff451090-inv_18DEC22045557-P1BS-059420282180_01_P005-BROWSE.jpg --profile csda-cumulus-sbx-7894
```
Example Delete using the second method
```
// Multiple files from a csv - Assuming the CSV is ONLY keypaths here, 1 column, and just the keypath in the column starting at the bucket's root dir
cat my_file_list.csv | xargs -P1 -n1000 bash -c 'aws s3api delete-objects --profile csda-cumulus-sbx-7894 --bucket deletebucket-test-sbx --delete "Objects=[$(printf "{Key=%s}," "$@")],Quiet=false"'
// To do a test run where nothing gets deleted, use the --dryrun parameter like this:
cat my_file_list.csv | xargs -P1 -n1000 bash -c 'aws s3api delete-objects --profile csda-cumulus-sbx-7894 --bucket deletebucket-test-sbx --delete "Objects=[$(printf "{Key=%s}," "$@")],Quiet=false"' --dryrun
// Also note, with this method, we were getting permission level errors until we moved the `--profile` parameter right up to be immediately after the `delete-objects` part of the command.
```

**Comment by krisstanton** [comment]
WIP UPDATE: 
Also wanted to mention that I disabled the inventories on `OLD_NGAP` and `NEW_CBA`.
These are inventories that I had previously setup.  At this point, the snapshots that we have are enough to complete this task.
Note: There is still an inventory on the public bucket called, `thumbnails` which is set to weekly that is still running on `OLD_NGAP__PROD_5982`.  I believe this is part of the PSScene3Band migration task.
Also note: When considering the Data Management System (DMS) read tasks, we need to consider the volume of files that are generated by these inventories.  It is something close to 60 files that are between 10 and 30 MB when compressed for each inventory that is generated.  That comes out to around 2 GB per inventory period generation (in plain english, that means that if we generate this inventory daily, then we have 2 Gigs per day being added)

**Comment by jsrikish** [comment]
Copied a few granules to a test bucket "deletebucket-test-sbx" in the sandbox account 7894
The fastest way to delete objects is using aws cli.  We have a list of files to be deleted saved in a csv file.    
To delete files listed in "delete.csv":
cat delete.csv | xargs -P8 -n1000 bash -c 'aws s3api --profile csda-cumulus-cba-7894 delete-objects  --bucket deletebucket-test-sbx --delete "Objects=[$(printf "{Key=%s}," "$@")],Quiet=false"' >& filesdeleted_in_{bucket}.out   

where -P is the maxprocs option
xargs to run up to this many processes simultaneously

-n is the max no. of arguments  to xargs

**Comment by krisstanton** [comment]
WIP Update:
	I've finished creating the Safe_To_Delete lists for all datasets (except for Planet - since we have not yet migrated Planet data)
	The outputs of key lists can be found on MCP at this directory:
		s3://csdap-maxar-delivery/utils/kris/OLD_NGAP_Deletes/2024-03-25/

We have completed removal of WV04_Pan and WV04_MSI from OLD_NGAP
The only remaining part of this overall task is to run the actual deletes for the remaining datasets.
Note: The remaining datasets are all significantly larger than the two WV04 datasets.
Using a laptop, we calculate that it would take multiple days of continuous operations to delete all of these files.
We will be running the actual bulk of the delete operations on the NCCS system for speed.

- [ ] Will update the File List and Size metrics counts after the deletes are confirmed.

I'm also working on uploading all the custom code tools I developed for this and intermediate files used to generate these final Safe_To_Delete lists.  Will post another WIP update when those file uploads are completed.
We will need to run this process again in the future when doing the MCP delete and possibly when we complete the planet scope migration as well.

**Comment by krisstanton** [comment]
WIP Update:
Updated the metrics and added a couple last items to finish up this ticket.

Here are the things that were added to the description above:

AFTER Almost all of the Deletes
- Old NGAP: csdap-cumulus-prod-protected
  - Total Bucket Size: 1.1 PB
  - Total Objects: 43938644 or (43,938,644) or ~43 million
- Old NGAP: csdap-cumulus-prod-public
  - Total Bucket Size: 73.2 GB
  - Total Objects: 589943 or (589,943) or ~0.5 million

AFTER all of the Deletes (including duplicate 2-digit and 1-digit directory name bug)
- Placeholder


- [ ] Remove the final duplicates (2 and 1 digit directory names) and update the metrics above again.
- [ ] Make an inventory so we know which granules were not fully ingested before shutting down NGAP server

**Comment by krisstanton** [comment]
Added some more detail to the inventory check box:

- [ ] Make an inventory so we know which granules were not fully ingested before shutting down NGAP server 
  - [ ] Turning the inventory back on for Protected and Public buckets for 1 day AFTER the above item is completed 
  - [ ] Download the latest inventory (and place the info in MCP Delivery bucket)
  - [ ] Turn that inventory back off

**Comment by krisstanton** [comment]
Final Numbers for this ticket (until we do migration of PSScene3Band)
	Old NGAP: csdap-cumulus-prod-protected
		Total Bucket Size: 877.6 TB
		Total Objects: 40647308 or (40,647,308) or ~40 million
	Old NGAP: csdap-cumulus-prod-public
		Total Bucket Size: 6.6 GB
		Total Objects: 41293 or (41,293) or ~41 thousand

	The majority of the remaining files are the PSScene3Band which have not yet been migration ingested.
	Since the PSScene3Band thumbs where never put in Old NGAP, We can use the public bucket object count as a near proxy to how many granules did not get migrated to New CBA across all Maxar Datasets.
	So that means that out of about 14 million total Maxar (WorldView and GeoEye) granules, only about 40k did not get migrated from Old NGAP to New CBA.
